oauthproxy.go

```diff
@@ -947,3 +975,75 @@ func (p *OAuthProxy) ErrorJSON(rw http.ResponseWriter, code int) {
 	rw.Header().Set("Content-Type", applicationJSON)
 	rw.WriteHeader(code)
 }
+
+// GetJwtSession loads a session based on a JWT token in the authorization header.
+func (p *OAuthProxy) GetJwtSession(req *http.Request) (*providers.SessionState, error) {
+	auth := req.Header.Get("Authorization")
+	ctx := context.Background()
+	var session *providers.SessionState
+
+	s := strings.SplitN(auth, " ", 2)
+	if len(s) != 2 {
+		return nil, fmt.Errorf("invalid Authorization header %s", req.Header.Get("Authorization"))
+	}
+
+	var rawBearerToken string
+	// Check if we have a Bearer token Masquerading as Basic
+	if s[0] == "Basic" {
+		b, err := b64.StdEncoding.DecodeString(s[1])
+		if err != nil {
+			return nil, err
+		}
+		pair := strings.SplitN(string(b), ":", 2)
+		if len(pair) != 2 {
+			return nil, fmt.Errorf("invalid format %s", b)
+		}
+		user, password := pair[0], pair[1]
+		if password == "x-oauth-basic" || password == "" {
+			rawBearerToken = user
+		} else if user == "x-oauth-basic" {
+			rawBearerToken = password
+		}
+	} else if s[0] == "Bearer" {
+		rawBearerToken = s[1]
+	} else {
+		return nil, fmt.Errorf("invalid Authorization header %s", req.Header.Get("Authorization"))
+	}
+
+	for _, verifier := range p.jwtBearerVerifiers {
+		bearerToken, err := verifier.Verify(ctx, rawBearerToken)
+
+		if err != nil {
+			log.Printf("failed to verify bearer token: %v", err)
+			continue
+		}
+
+		var claims struct {
+			Email    string `json:"email"`
+			Verified *bool  `json:"email_verified"`
+		}
+
+		if err := bearerToken.Claims(&claims); err != nil {
+			return nil, fmt.Errorf("failed to parse bearer token claims: %v", err)
+		}
+
+		if claims.Email == "" {
+			return nil, fmt.Errorf("id_token did not contain an email")
```

Comment:
You might want to check what https://github.com/pusher/oauth2_proxy/pull/57 is doing in this case, perhaps there is some way we can deduplicate this code

Reply:
Implemented what was in #57 - if `claims.Email == ""`, then `claims.Email = claims.Subject`